docs/root/start/sandboxes/single-page-app.rst

```diff
@@ -0,0 +1,603 @@
+    single-page-app-envoy-1       single-page-app-envoy       "/docker-entrypoint.sh envoy -c /etc/envoy/envoy.yaml ..." envoy     2 minutes ago   Up 2 minutes           0.0.0.0:10000-10001->10000-10001/tcp, :::10000-10001->10000-10001/tcp
+    single-page-app-myhub-1       single-page-app-myhub       "/opt/myhub/app.py"                                        myhub     2 minutes ago   Up 2 minutes (healthy) 0.0.0.0:7000->7000/tcp, :::7000->7000/tcp
+    single-page-app-myhub-api-1   single-page-app-myhub-api   "/opt/myhub/app.py"                                        myhub-api 2 minutes ago   Up 2 minutes (healthy)
+    single-page-app-ui-1          single-page-app-ui          "/entrypoint.sh dev.sh"                                    ui        2 minutes ago   Up 2 minutes (healthy)
+
+.. _install_sandboxes_single_page_app_step_login:
+
+Step 3: Browse to the dev app and login
+***************************************
+
+The development app should now be available at http://localhost:10001 and provide a login button:
+
+.. image:: /start/sandboxes/_static/spa-login.png
+   :align: center
+
+.. note::
+   The dummy OAuth provider automatically trusts everyone as a hard-coded ``envoydemo`` user and redirects back to the app.
+
+   In a real world scenario the provider would authenticate and authorize the user before proceeding.
+
+The sandbox is configured with an inverted match on
+:ref:`pass_through_matcher <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.pass_through_matcher>`.
+
+This ignores all paths for OAuth other than:
+
+- ``/authorize.*``
+- ``/hub.*``
+- ``/login``
+- ``/logout``.
+
+.. literalinclude:: _include/single-page-app/envoy.yml
+   :language: yaml
+   :lines: 37-46
+   :linenos:
+   :lineno-start: 37
+   :emphasize-lines: 3-8
+   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`
+
+When a user clicks ``login`` the app initiates the OAuth flow by calling the ``/login`` path in Envoy.
+
+This redirects the user to the OAuth provider for authorization/authentication with a further redirect link:
+
+.. literalinclude:: _include/single-page-app/envoy.yml
+   :language: yaml
+   :lines: 34-39
+   :linenos:
+   :lineno-start: 34
+   :emphasize-lines: 3-4
+   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`
+
+On successful authorization/authentication the user is redirected back via this link to the app with the necessary OAuth
+`authorization code <oauth-auth-code_>`__ to proceed:
+
+.. literalinclude:: _include/single-page-app/envoy.yml
+   :language: yaml
+   :lines: 44-50
+   :linenos:
+   :lineno-start: 44
+   :emphasize-lines: 3-5
+   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`
+
+Envoy then uses this authorization code with its client secret to generate an access token for the user, which Envoy then retains:
```

Comment:
I don't think it retains it. I think it turns it into a client cookie, see https://github.com/envoyproxy/envoy/blob/12210e53db6cea79b58ee027f047faa356e1e392/source/extensions/filters/http/oauth2/filter.cc#L565.

Reply:
im pretty sure envoy keeps something - if i delete my cooking and refresh the page it reappears - again i should probs double check this - i was basing this of what i observed not checking source

Reply:
k - checking this more throroughly - it doesnt retain anything afaict

i tracked down why this seemed to be the case - i was using http://localhost:10001/hub/user directly to check credentials - but this endpoint triggers the oauth flow so i was simply reauthing

ill update (still not clear what the hmac used for but its not for storing anything in envoy afaict)